OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs

```diff
@@ -0,0 +1,224 @@
+		readonly ModContent.ModDownload download;
+		readonly Action onSuccess;
+
+		readonly Widget panel;
+		readonly ProgressBarWidget progressBar;
+
+		Func<string> getStatusText = () => "";
+		string downloadHost;
+
+		[ObjectCreator.UseCtor]
+		public DownloadPackageLogic(Widget widget, ModContent.ModDownload download, Action onSuccess)
+		{
+			this.download = download;
+			this.onSuccess = onSuccess;
+
+			Log.AddChannel("install", "install.log");
+
+			panel = widget.Get("PACKAGE_DOWNLOAD_PANEL");
+			progressBar = panel.Get<ProgressBarWidget>("PROGRESS_BAR");
+
+			var statusLabel = panel.Get<LabelWidget>("STATUS_LABEL");
+			var statusFont = Game.Renderer.Fonts[statusLabel.Font];
+			var status = new CachedTransform<string, string>(s => WidgetUtils.TruncateText(s, statusLabel.Bounds.Width, statusFont));
+			statusLabel.GetText = () => status.Update(getStatusText());
+
+			var text = "Downloading {0}".F(download.Title);
+			panel.Get<LabelWidget>("TITLE").Text = text;
+
+			ShowDownloadDialog();
+		}
+
+		void ShowDownloadDialog()
+		{
+			getStatusText = () => "Fetching list of mirrors...";
+			progressBar.Indeterminate = true;
+
+			var retryButton = panel.Get<ButtonWidget>("RETRY_BUTTON");
+			retryButton.IsVisible = () => false;
+
+			var cancelButton = panel.Get<ButtonWidget>("CANCEL_BUTTON");
+
+			var file = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+
+			Action<DownloadProgressChangedEventArgs> onDownloadProgress = i =>
+			{
+				var dataReceived = 0.0f;
+				var dataTotal = 0.0f;
+				var mag = 0;
+				var dataSuffix = "";
+
+				if (i.TotalBytesToReceive < 0)
+				{
+					dataTotal = float.NaN;
+					dataReceived = i.BytesReceived;
+					dataSuffix = SizeSuffixes[0];
+				}
+				else
+				{
+					mag = (int)Math.Log(i.TotalBytesToReceive, 1024);
+					dataTotal = i.TotalBytesToReceive / (float)(1L << (mag * 10));
+					dataReceived = i.BytesReceived / (float)(1L << (mag * 10));
+					dataSuffix = SizeSuffixes[mag];
+				}
+
+				progressBar.Indeterminate = false;
+				progressBar.Percentage = i.ProgressPercentage;
+
+				getStatusText = () => "Downloading from {4} {1:0.00}/{2:0.00} {3} ({0}%)".F(i.ProgressPercentage,
+					dataReceived, dataTotal, dataSuffix,
+					downloadHost ?? "unknown host");
+			};
+
+			Action<string> onExtractProgress = s => Game.RunAfterTick(() => getStatusText = () => s);
+
+			Action<string> onError = s => Game.RunAfterTick(() =>
+			{
+				Log.Write("install", "Download failed: " + s);
+
+				progressBar.Indeterminate = false;
+				progressBar.Percentage = 100;
+				getStatusText = () => "Error: " + s;
+				retryButton.IsVisible = () => true;
+			});
+
+			Action<AsyncCompletedEventArgs, bool> onDownloadComplete = (i, cancelled) =>
+			{
+				if (i.Error != null)
+				{
+					onError(Download.FormatErrorMessage(i.Error));
+					return;
+				}
+
+				if (cancelled)
+				{
+					onError("Download cancelled");
+					return;
+				}
+
+				// Automatically extract
```

Comment:
What if I don't want the download to be extracted, but just copied straight to Contents/mymod/? Is that an unreasonable use case? Afaik, we support .zip as archive for game files, and I doubt most people would go to the trouble of creating .mix files.